From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Martijn Muijsers <martijnmuijsers@live.nl>
Date: Wed, 10 Jan 2024 19:37:04 +0100
Subject: [PATCH] Modifiable Bukkit enums - Material - Expand matching

Expands and generalizes the implementation of the `Material.matchMaterial` method in 2 ways:
* Matching is attempted strictly (case-sensitive) first, then case-insensitive, and then with leaving out increasingly lesser important characters.
* Matching is performed on the namespaced key, the namespaced key path and the enum name.

This means plugins that use `matchMaterial` to find the `Material` for a string will always find it, whether the string they have
(that occurred in a database, was given by a player, etc.) was originally based on the namespaced key, the enum name, and whether or not
it is complete, with the correct case, etc.

License: GPL-3.0 (https://www.gnu.org/licenses/gpl-3.0.html)
Fiddle - https://fiddlemc.org

diff --git a/src/main/java/org/bukkit/Material.java b/src/main/java/org/bukkit/Material.java
index 26438928078b25496003ede8231c8ca6624b5112..9f86d681430f53feff3de85824929cf7d80f7325 100644
--- a/src/main/java/org/bukkit/Material.java
+++ b/src/main/java/org/bukkit/Material.java
@@ -5022,11 +5022,30 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
     /**
      * Attempts to match the Material with the given name.
      * <p>
-     * This is a match lookup; names will be stripped of the "minecraft:"
-     * namespace, converted to uppercase, then stripped of special characters in
-     * an attempt to format it like the enum.
+     * <!-- // Fiddle start - modifiable Bukkit enums - expand Material matching -->
+     * This is a match lookup.
+     * It will attempt to match the given the {@code string} to a {@link Material},
+     * by comparing it to some string identifying that Material.
+     * <br>
+     * The following comparisons are performed, in the given order:
+     * <ol>
+     *     <li>An exact match</li>
+     *     <li>A match ignoring case</li>
+     *     <li>A match ignoring case and only checking letters, digits, <code>_</code> and <code>:</code></li>
+     *     <li>A match ignoring case and only checking letters, digits and <code>_</code></li>
+     *     <li>A match ignoring case and only checking letters and digits</li>
+     * </ol>
+     * For each of the above comparisons, the following string will be checked against for a certain Material:
+     * <ol>
+     *     <li>The {@linkplain #getKey() namespaced key}, for example: <code>"minecraft:stone_bricks"</code> or <code>"masonry:tuff_bricks"</code></li>
+     *     <li>The namespaced key path (part behind the : in a namespaced key), for example: <code>"stone_bricks"</code> or <code>"tuff_bricks"</code></li>
+     *     <li>The enum name, for example: <code>"STONE_BRICKS"</code></li>
+     * </ol>
+     * In other words, first the given string will be checked case-sensitively against the namespaced key,
+     * then case-sensitively against the namespaced key path, and so on.
      *
-     * @param name Name of the material to get
+     * @param name String of the material to get
+     * <!-- // Fiddle end - modifiable Bukkit enums - expand Material matching -->
      * @return Material if found, or null
      */
     @Nullable
@@ -5036,15 +5055,23 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
 
     /**
      * Attempts to match the Material with the given name.
-     * <p>
-     * This is a match lookup; names will be stripped of the "minecraft:"
-     * namespace, converted to uppercase, then stripped of special characters in
-     * an attempt to format it like the enum.
-     *
-     * @param name Name of the material to get
+     * <!-- // Fiddle start - modifiable Bukkit enums - expand Material matching -->
+     * Returns the same as {@link #matchMaterial(String)} if {@code legacyName} is false.
+     * If {@code legacyName} is true, the following steps are performed on the string in order:
+     * <ol>
+     *     <li>If it starts with the case-sensitive prefix <code>"minecraft:"</code>, that prefix is removed.</li>
+     *     <li>It is transformed to uppercase.</li>
+     *     <li>Every contiguous substring of whitespace characters (as given by regex <code>\s</code>) is replaced by a single underscore (<code>_</code>).</li>
+     *     <li>Every non-word character (as given by regex <code>\W</code>) is removed.</li>
+     *     <li>If it does not start with the prefix {@link org.fiddlemc.fiddle.material.MaterialLegacyNameChecker#LEGACY_MATERIAL_PREFIX}, that prefix is prepended.</li>
+     *     <li>Finally, a {@link Material} with exactly the string as its enum name, if present, is mapped to its non-legacy variant (with priority for matching blocks over items) and returned.</li>
+     * </ol>
+     * @param name String of the material to get
+     * <!-- // Fiddle end - modifiable Bukkit enums - expand Material matching -->
      * @param legacyName whether this is a legacy name (see
      * {@link #getMaterial(java.lang.String, boolean)}
      * @return Material if found, or null
+     * @see #matchMaterial(String) <!-- // Fiddle - modifiable Bukkit enums - expand Material matching -->
      */
     @Nullable
     public static Material matchMaterial(@NotNull final String name, boolean legacyName) {
diff --git a/src/main/java/org/fiddlemc/fiddle/material/MaterialNameAndKeyConversion.java b/src/main/java/org/fiddlemc/fiddle/material/MaterialNameAndKeyConversion.java
index 3c992c62764cb54e36a0409dd922dc48aaef1928..5a72421d3e60d493f0f2b6a0a3f75c078476d215 100644
--- a/src/main/java/org/fiddlemc/fiddle/material/MaterialNameAndKeyConversion.java
+++ b/src/main/java/org/fiddlemc/fiddle/material/MaterialNameAndKeyConversion.java
@@ -43,6 +43,121 @@ public final class MaterialNameAndKeyConversion {
     }
     // Fiddle end - modifiable Bukkit enums - Material - remove hard-coded properties - getMaterial
 
+    // Fiddle start - modifiable Bukkit enums - expand Material matching
+    private static @org.jetbrains.annotations.Nullable java.util.Map<String, Material> cachedMaterialsByMatchableString = null;
+
+    private static java.util.Map<String, Material> getCachedMaterialsByMatchableString() {
+        if (cachedMaterialsByMatchableString == null) {
+            cachedMaterialsByMatchableString = com.google.common.collect.Maps.newHashMap();
+            for (int valueIndex = 0; valueIndex < numberOfMatchMaterialValues; valueIndex++) {
+                for (int normalizationIndex = 0; normalizationIndex < NormalizedMatchMaterialStrings.numberOfNormalizations; normalizationIndex++) {
+                    for (var material : CachedMaterialProperty.getCachedMaterialValues()) {
+                        var value = getMatchMaterialValue(material, valueIndex);
+                        if (value != null) {
+                            var normalizedValue = new NormalizedMatchMaterialStrings(value).getNormalized(normalizationIndex);
+                            cachedMaterialsByMatchableString.putIfAbsent(normalizedValue, material);
+                        }
+                    }
+                }
+            }
+        }
+        return cachedMaterialsByMatchableString;
+    }
+
+    private static final int numberOfMatchMaterialValues = 3;
+
+    private static @org.jetbrains.annotations.Nullable String getMatchMaterialValue(Material material, int valueIndex) {
+        return switch (valueIndex) {
+            case 0 -> material.isLegacy() ? null : material.getKey().toString();
+            case 1 -> material.isLegacy() ? null : material.getKey().getKey();
+            case 2 -> material.name();
+            default -> throw new IllegalArgumentException("Invalid valueIndex: " + valueIndex);
+        };
+    }
+
+    private static class NormalizedMatchMaterialStrings {
+
+        public static final int numberOfNormalizations = 5;
+
+        public final String original;
+        private @org.jetbrains.annotations.Nullable String lowerCase;
+        private @org.jetbrains.annotations.Nullable String lowerCaseLettersDigitsUnderscoresAndColons;
+        private @org.jetbrains.annotations.Nullable String lowerCaseLettersDigitsAndUnderscores;
+        private @org.jetbrains.annotations.Nullable String lowerCaseLettersAndDigits;
+
+        public NormalizedMatchMaterialStrings(@org.jetbrains.annotations.NotNull String original) {
+            this.original = original;
+        }
+
+        public @org.jetbrains.annotations.NotNull String getLowerCase() {
+            if (this.lowerCase == null) {
+                this.lowerCase = this.original.toLowerCase(java.util.Locale.ROOT);
+            }
+            return this.lowerCase;
+        }
+
+        public @org.jetbrains.annotations.NotNull String getLowerCaseLettersDigitsUnderscoresAndColons() {
+            if (this.lowerCaseLettersDigitsUnderscoresAndColons == null) {
+                char[] characters = this.getLowerCase().toCharArray();
+                int writeIndex = 0;
+                for (int readIndex = 0; readIndex < characters.length; readIndex++) {
+                    char character = characters[readIndex];
+                    if ((character >= 'a' && character <= 'z') || (character >= '0' && character <= '9') || character == '_' || character == ':') {
+                        characters[writeIndex] = character;
+                        writeIndex++;
+                    }
+                }
+                this.lowerCaseLettersDigitsUnderscoresAndColons = new String(characters, 0, writeIndex);
+            }
+            return this.lowerCaseLettersDigitsUnderscoresAndColons;
+        }
+
+        public @org.jetbrains.annotations.NotNull String getLowerCaseLettersDigitsAndUnderscores() {
+            if (this.lowerCaseLettersDigitsAndUnderscores == null) {
+                char[] characters = this.getLowerCaseLettersDigitsUnderscoresAndColons().toCharArray();
+                int writeIndex = 0;
+                for (int readIndex = 0; readIndex < characters.length; readIndex++) {
+                    char character = characters[readIndex];
+                    if ((character >= 'a' && character <= 'z') || (character >= '0' && character <= '9') || character == '_') {
+                        characters[writeIndex] = character;
+                        writeIndex++;
+                    }
+                }
+                this.lowerCaseLettersDigitsAndUnderscores = new String(characters, 0, writeIndex);
+            }
+            return this.lowerCaseLettersDigitsAndUnderscores;
+        }
+
+        public @org.jetbrains.annotations.NotNull String getLowerCaseLettersAndDigits() {
+            if (this.lowerCaseLettersAndDigits == null) {
+                char[] characters = this.getLowerCaseLettersDigitsAndUnderscores().toCharArray();
+                int writeIndex = 0;
+                for (int readIndex = 0; readIndex < characters.length; readIndex++) {
+                    char character = characters[readIndex];
+                    if ((character >= 'a' && character <= 'z') || (character >= '0' && character <= '9')) {
+                        characters[writeIndex] = character;
+                        writeIndex++;
+                    }
+                }
+                this.lowerCaseLettersAndDigits = new String(characters, 0, writeIndex);
+            }
+            return this.lowerCaseLettersAndDigits;
+        }
+
+        public @org.jetbrains.annotations.NotNull String getNormalized(int normalizationIndex) {
+            return switch (normalizationIndex) {
+                case 0 -> this.original;
+                case 1 -> this.getLowerCase();
+                case 2 -> this.getLowerCaseLettersDigitsUnderscoresAndColons();
+                case 3 -> this.getLowerCaseLettersDigitsAndUnderscores();
+                case 4 -> this.getLowerCaseLettersAndDigits();
+                default -> throw new IllegalArgumentException("Invalid normalizationIndex: " + normalizationIndex);
+            };
+        }
+
+    }
+    // Fiddle end - modifiable Bukkit enums - expand Material matching
+
     // Fiddle start - modifiable Bukkit enums - Material - remove hard-coded properties - key
     /**
      * @throws IllegalArgumentException If the given {@code material} is {@linkplain Material#isLegacy() legacy}.
@@ -86,7 +201,20 @@ public final class MaterialNameAndKeyConversion {
      */
     public static @org.jetbrains.annotations.Nullable Material matchMaterial(final String name, boolean legacyName) {
         // Based on the original org.bukkit.Material#matchMaterial(String, boolean)
+        // Fiddle start - modifiable Bukkit enums - expand Material matching
         com.google.common.base.Preconditions.checkArgument(name != null, "Name cannot be null");
+        if (!legacyName) {
+            var cachedMaterialsByMatchableString = getCachedMaterialsByMatchableString();
+            var normalizedStrings = new NormalizedMatchMaterialStrings(name);
+            for (int normalizationIndex = 0; normalizationIndex < NormalizedMatchMaterialStrings.numberOfNormalizations; normalizationIndex++) {
+                var materialFound = cachedMaterialsByMatchableString.get(normalizedStrings.getNormalized(normalizationIndex));
+                if (materialFound != null) {
+                    return materialFound;
+                }
+            }
+            return null;
+        }
+        // Fiddle end - modifiable Bukkit enums - expand Material matching
 
         String filtered = name;
         if (filtered.startsWith(org.bukkit.NamespacedKey.MINECRAFT + ":")) {
